rcl/include/rcl/subscription.h

```diff
@@ -0,0 +1,266 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__SUBSCRIPTION_H_
+#define RCL__SUBSCRIPTION_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rosidl_generator_c/message_type_support.h"
+
+#include "rcl/node.h"
+
+/// Internal rcl implementation struct.
+struct rcl_subscription_impl_t;
+
+/// Handle for a rcl subscription.
+typedef struct rcl_subscription_t {
+  struct rcl_subscription_impl_t * impl;
+} rcl_subscription_t;
+
+/// Options available for a rcl subscription.
+typedef struct rcl_subscription_options_t {
+  /// Middleware quality of service settings for the subscription.
+  rmw_qos_profile_t qos;
+  /// If true, messages published from within the same node are ignored.
+  bool ignore_local_publications;
+  /// Custom allocator for the subscription, used for incidental allocations.
+  /* For default behavior (malloc/free), see: rcl_get_default_allocator() */
+  rcl_allocator_t allocator;
+} rcl_subscription_options_t;
+
+/// Return a rcl_subscription_t struct with members set to NULL.
+/* Should be called to get a null rcl_subscription_t before passing to
+ * rcl_initalize_subscription().
+ * It's also possible to use calloc() instead of this if the rcl_subscription_t
+ * is being allocated on the heap.
+ */
+rcl_subscription_t
+rcl_get_zero_initialized_subscription();
+
+/// Initialize a ROS subscription.
+/* After calling this function on a rcl_subscription_t, it can be used to take
+ * messages of the given type to the given topic using rcl_take().
+ *
+ * The given rcl_node_t must be valid and the resulting rcl_subscription_t is
+ * only valid as long as the given rcl_node_t remains valid.
+ *
+ * The rosidl_message_type_support_t is obtained on a per .msg type basis.
+ * When the user defines a ROS message, code is generated which provides the
+ * required rosidl_message_type_support_t object.
+ * This object can be obtained using a language appropriate mechanism.
+ * \TODO(wjwwood) probably should talk about this once and link to it instead
```

Comment:
agreed; you can use Doxygen \sa to reference a different function and file

https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdsa